File: Sources/AgentCoding/Mitm/SigV4Signer.swift
import Foundation
import Crypto

/// Pure SigV4 signing primitives — no HTTP I/O. `AWSResigner` calls
/// `sign(...)` to compute fresh signing material from credentials that
/// only ever live in this process's address space.
///
/// References:
///   - https://docs.aws.amazon.com/general/latest/gr/sigv4_signing.html
///   - Test vectors: https://docs.aws.amazon.com/general/latest/gr/signature-v4-test-suite.html
///
/// The "get-vanilla" reference vector is reproduced step-by-step in
/// the comments at the bottom of this file so the canonical-request
/// shape, key-derivation chain, and final signature can be checked
/// by inspection against an AWS-published expected.
public enum SigV4Signer {
    public struct Credentials: Sendable {
        public let accessKeyID: String
        public let secretAccessKey: String
        /// Optional STS session token. When present the caller MUST
        /// also include it as the `X-Amz-Security-Token` header in
        /// `Request.headers` so it gets folded into canonical-headers.
        public let sessionToken: String?
        public init(accessKeyID: String,
                    secretAccessKey: String,
                    sessionToken: String? = nil) {
            self.accessKeyID = accessKeyID
            self.secretAccessKey = secretAccessKey
            self.sessionToken = sessionToken
        }
    }

    public struct Scope: Sendable {
        /// `YYYYMMDD` portion of `X-Amz-Date`.
        public let date: String
        public let region: String
        public let service: String
        public init(date: String, region: String, service: String) {
            self.date = date
            self.region = region
            self.service = service
        }
        public var credentialScope: String {
            "\(date)/\(region)/\(service)/aws4_request"
        }
    }

    public struct Request {
        public let method: String
        /// Already-encoded path component (everything from the request
        /// line up to `?`). The SDK percent-encodes per the service's
        /// rules; we forward whatever the SDK produced.
        public let path: String
        /// Already-encoded query string (no leading `?`), in **wire**
        /// order. The signer canonicalizes (sort + join) internally —
        /// callers don't need to pre-sort.
        public let query: String
        /// Headers as `(name, value)` in submission order. Names case-
        /// insensitive; duplicates allowed (combined per RFC during
        /// canonicalization).
        public let headers: [(String, String)]
        public let body: Data
        public init(method: String, path: String, query: String,
                    headers: [(String, String)], body: Data) {
            self.method = method
            self.path = path
            self.query = query
            self.headers = headers
            self.body = body
        }
    }

    public struct SignedOutput {
        /// `AWS4-HMAC-SHA256 Credential=…, SignedHeaders=…, Signature=…`
        public let authorization: String
        /// Hex SHA256 of the body — or `UNSIGNED-PAYLOAD` when the
        /// caller passed `unsignedPayload: true`.
        public let payloadHash: String
        /// `YYYYMMDDTHHMMSSZ` — the `X-Amz-Date` header value.
        public let amzDate: String
        /// Sorted, semicolon-joined lowercase header names.
        public let signedHeaders: String
    }

    /// Sign the request. Caller must already have the headers list it
    /// intends to put on the wire (Host, Content-Type, X-Amz-Date,
    /// X-Amz-Content-SHA256, X-Amz-Security-Token if applicable, …) —
    /// `Authorization` itself is the only header excluded.
    public static func sign(
        request: Request,
        credentials: Credentials,
        scope: Scope,
        date: Date = Date(),
        unsignedPayload: Bool = false
    ) -> SignedOutput {
        let amzDate = isoBasic(date)

        let payloadHash: String
        if unsignedPayload {
            payloadHash = "UNSIGNED-PAYLOAD"
        } else {
            payloadHash = hexSHA256(request.body)
        }

        // Canonicalize headers. Values get whitespace-trimmed; runs of
        // internal whitespace outside of double-quoted strings are
        // collapsed to a single space. Names lowercased and sorted.
        // Duplicates joined with ',' (RFC 9110 §5.3).
        var byName: [String: [String]] = [:]
        for (k, v) in request.headers {
            let key = k.lowercased()
            byName[key, default: []].append(canonicalHeaderValue(v))
        }
        let names = byName.keys.sorted()
        var canonicalHeaders = ""
        for n in names {
            canonicalHeaders += "\(n):\(byName[n]!.joined(separator: ","))\n"
        }
        let signedHeaders = names.joined(separator: ";")

        let canonicalRequest = [
            request.method.uppercased(),
            request.path.isEmpty ? "/" : request.path,
            canonicalQueryString(request.query),
            canonicalHeaders,
            signedHeaders,
            payloadHash
        ].joined(separator: "\n")

        let stringToSign = [
            "AWS4-HMAC-SHA256",
            amzDate,
            scope.credentialScope,
            hexSHA256(Data(canonicalRequest.utf8))
        ].joined(separator: "\n")

        // Key-derivation chain: kSecret → kDate → kRegion → kService → kSigning.
        let kSecret  = Data("AWS4\(credentials.secretAccessKey)".utf8)
        let kDate    = hmac(key: kSecret,  msg: Data(scope.date.utf8))
        let kRegion  = hmac(key: kDate,    msg: Data(scope.region.utf8))
        let kService = hmac(key: kRegion,  msg: Data(scope.service.utf8))
        let kSigning = hmac(key: kService, msg: Data("aws4_request".utf8))

        let signature = hmac(key: kSigning, msg: Data(stringToSign.utf8))
            .map { String(format: "%02x", $0) }.joined()

        let auth = "AWS4-HMAC-SHA256 " +
            "Credential=\(credentials.accessKeyID)/\(scope.credentialScope), " +
            "SignedHeaders=\(signedHeaders), " +
            "Signature=\(signature)"

        return SignedOutput(authorization: auth,
                            payloadHash: payloadHash,
                            amzDate: amzDate,
                            signedHeaders: signedHeaders)
    }

    // MARK: - helpers

    public static func hexSHA256(_ data: Data) -> String {
        SHA256.hash(data: data).map { String(format: "%02x", $0) }.joined()
    }

    private static func hmac(key: Data, msg: Data) -> Data {
        let symKey = SymmetricKey(data: key)
        let mac = HMAC<SHA256>.authenticationCode(for: msg, using: symKey)
        return Data(mac)
    }

    /// `YYYYMMDDTHHMMSSZ` (ISO 8601 basic) in UTC.
    public static func isoBasic(_ date: Date) -> String {
        let f = DateFormatter()
        f.timeZone = TimeZone(identifier: "UTC")
        f.locale = Locale(identifier: "en_US_POSIX")
        f.dateFormat = "yyyyMMdd'T'HHmmss'Z'"
        return f.string(from: date)
    }

    /// Canonical query string per SigV4: split on `&`, sort by key
    /// (then by value), join `key=value` with `&`. Caller is expected
    /// to have URI-encoded keys/values per the service's rules — the
    /// SDK does this for any request it intends to sign.
    static func canonicalQueryString(_ query: String) -> String {
        if query.isEmpty { return "" }
        let pairs = query.split(separator: "&", omittingEmptySubsequences: false)
            .map(String.init)
        let parsed: [(String, String)] = pairs.map { p in
            if let eq = p.firstIndex(of: "=") {
                return (String(p[..<eq]), String(p[p.index(after: eq)...]))
            }
            return (p, "")
        }
        let sorted = parsed.sorted { a, b in
            if a.0 != b.0 { return a.0 < b.0 }
            return a.1 < b.1
        }
        return sorted.map { "\($0.0)=\($0.1)" }.joined(separator: "&")
    }

    /// Trim leading/trailing whitespace and collapse runs of internal
    /// whitespace (outside of double-quoted strings) to a single space.
    /// Per the SigV4 canonical-headers rule.
    static func canonicalHeaderValue(_ s: String) -> String {
        let trimmed = s.trimmingCharacters(in: .whitespaces)
        var out = ""
        var inQuotes = false
        var lastWasSpace = false
        for c in trimmed {
            if c == "\"" { inQuotes.toggle(); out.append(c); lastWasSpace = false; continue }
            if !inQuotes, c == " " || c == "\t" {
                if !lastWasSpace { out.append(" "); lastWasSpace = true }
            } else {
                out.append(c); lastWasSpace = false
            }
        }
        return out
    }
}

// MARK: - Reference vector documentation
//
// AWS "get-vanilla" test from the SigV4 test suite (paraphrased):
//
//   Method: GET
//   Path:   /
//   Query:  (empty)
//   Headers:
//     Host:        example.amazonaws.com
//     X-Amz-Date:  20150830T123600Z
//   Body: (empty)
//
//   AccessKey: AKIDEXAMPLE
//   Secret:    wJalrXUtnFEMI/K7MDENG+bPxRfiCYEXAMPLEKEY
//   Region:    us-east-1
//   Service:   service
//   Date:      20150830T123600Z
//
// Expected canonical request:
//   GET\n
//   /\n
//   \n
//   host:example.amazonaws.com\n
//   x-amz-date:20150830T123600Z\n
//   \n
//   host;x-amz-date\n
//   e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
//
// Expected string-to-sign:
//   AWS4-HMAC-SHA256\n
//   20150830T123600Z\n
//   20150830/us-east-1/service/aws4_request\n
//   bb579772317eb040ac9ed261061d46c1f17a8133879d6129b6e1c25292927e63
//
// Expected signature:
//   5fa00fa31553b73ebf1942676e86291e8372ff2a2260956d9b8aae1d763fbf31
//
// Expected Authorization:
//   AWS4-HMAC-SHA256
//   Credential=AKIDEXAMPLE/20150830/us-east-1/service/aws4_request,
//   SignedHeaders=host;x-amz-date,
//   Signature=5fa00fa31553b73ebf1942676e86291e8372ff2a2260956d9b8aae1d763fbf31

File: windows/Bromure.AC.Mitm/SigV4/SigV4Signer.cs
// macos-source: Sources/AgentCoding/Mitm/SigV4Signer.swift @ d9041f7dd0b1
using System.Globalization;
using System.Security.Cryptography;
using System.Text;

namespace Bromure.AC.Mitm.SigV4;

/// <summary>
/// Direct port of <c>Sources/AgentCoding/Mitm/SigV4Signer.swift</c>.
/// Pure SigV4 signing primitives — no HTTP I/O. <c>AwsResigner</c> calls
/// <see cref="Sign"/> to compute fresh signing material from credentials
/// that only ever live in this process's address space.
///
/// <para>References:
/// <list type="bullet">
///   <item>https://docs.aws.amazon.com/general/latest/gr/sigv4_signing.html</item>
///   <item>https://docs.aws.amazon.com/general/latest/gr/signature-v4-test-suite.html</item>
/// </list>
/// </para>
///
/// The "get-vanilla" reference vector is reproduced step-by-step in the
/// trailing comments so the canonical-request shape, key-derivation
/// chain, and final signature can be checked by inspection. Mirrors the
/// vector cited in the macOS source.
/// </summary>
public static class SigV4Signer
{
    public sealed record Credentials(string AccessKeyId, string SecretAccessKey, string? SessionToken = null);

    public sealed record Scope(string Date, string Region, string Service)
    {
        public string CredentialScope => $"{Date}/{Region}/{Service}/aws4_request";
    }

    public sealed record Request(
        string Method,
        string Path,
        string Query,
        IReadOnlyList<(string Name, string Value)> Headers,
        ReadOnlyMemory<byte> Body);

    public sealed record SignedOutput(
        string Authorization,
        string PayloadHash,
        string AmzDate,
        string SignedHeaders);

    public static SignedOutput Sign(
        Request request,
        Credentials credentials,
        Scope scope,
        DateTime? date = null,
        bool unsignedPayload = false)
    {
        var amzDate = IsoBasic(date ?? DateTime.UtcNow);

        var payloadHash = unsignedPayload ? "UNSIGNED-PAYLOAD" : HexSha256(request.Body.Span);

        // Canonicalize headers. Values get whitespace-trimmed; runs of
        // internal whitespace outside of double-quoted strings are
        // collapsed to a single space. Names lowercased and sorted.
        var byName = new SortedDictionary<string, List<string>>(StringComparer.Ordinal);
        foreach (var (k, v) in request.Headers)
        {
            var key = k.ToLowerInvariant();
            if (!byName.TryGetValue(key, out var list))
            {
                list = new List<string>();
                byName[key] = list;
            }
            list.Add(CanonicalHeaderValue(v));
        }
        var sb = new StringBuilder();
        foreach (var (n, vs) in byName)
        {
            sb.Append(n).Append(':').Append(string.Join(',', vs)).Append('\n');
        }
        var canonicalHeaders = sb.ToString();
        var signedHeaders = string.Join(';', byName.Keys);

        var canonicalRequest = string.Join('\n',
            request.Method.ToUpperInvariant(),
            CanonicalPath(request.Path, scope.Service),
            CanonicalQueryString(request.Query),
            canonicalHeaders,
            signedHeaders,
            payloadHash);

        var stringToSign = string.Join('\n',
            "AWS4-HMAC-SHA256",
            amzDate,
            scope.CredentialScope,
            HexSha256(Encoding.UTF8.GetBytes(canonicalRequest)));

        var kSecret = Encoding.UTF8.GetBytes("AWS4" + credentials.SecretAccessKey);
        var kDate = Hmac(kSecret, Encoding.UTF8.GetBytes(scope.Date));
        var kRegion = Hmac(kDate, Encoding.UTF8.GetBytes(scope.Region));
        var kService = Hmac(kRegion, Encoding.UTF8.GetBytes(scope.Service));
        var kSigning = Hmac(kService, Encoding.UTF8.GetBytes("aws4_request"));

        var signature = ToHex(Hmac(kSigning, Encoding.UTF8.GetBytes(stringToSign)));

        var authorization = "AWS4-HMAC-SHA256 "
            + $"Credential={credentials.AccessKeyId}/{scope.CredentialScope}, "
            + $"SignedHeaders={signedHeaders}, "
            + $"Signature={signature}";

        return new SignedOutput(authorization, payloadHash, amzDate, signedHeaders);
    }

    public static string HexSha256(ReadOnlySpan<byte> data)
    {
        Span<byte> hash = stackalloc byte[32];
        SHA256.HashData(data, hash);
        return ToHex(hash);
    }

    private static byte[] Hmac(byte[] key, byte[] msg) => HMACSHA256.HashData(key, msg);

    private static string ToHex(ReadOnlySpan<byte> data)
    {
        var sb = new StringBuilder(data.Length * 2);
        foreach (var b in data) sb.Append(b.ToString("x2", CultureInfo.InvariantCulture));
        return sb.ToString();
    }

    /// <summary><c>YYYYMMDDTHHMMSSZ</c> in UTC.</summary>
    public static string IsoBasic(DateTime date)
    {
        var d = date.Kind == DateTimeKind.Utc ? date : date.ToUniversalTime();
        return d.ToString("yyyyMMdd'T'HHmmss'Z'", CultureInfo.InvariantCulture);
    }

    /// <summary>
    /// Canonical path per SigV4. For most services the wire path is
    /// URI-encoded once for the canonical request (so a literal <c>:</c>
    /// becomes <c>%3A</c>, and a wire <c>%20</c> becomes <c>%2520</c>).
    /// S3 is the documented exception that passes the path through
    /// untouched.
    ///
    /// <para>Bedrock model IDs (<c>global.anthropic.claude-sonnet-4-5-…-v1:0</c>)
    /// always carry a <c>:</c>, which is exactly the kind of char this
    /// rule exists for — without re-encoding, every Bedrock call signs
    /// against a canonical that diverges from AWS's by one byte per
    /// <c>:</c> and AWS rejects with InvalidSignatureException.</para>
    /// </summary>
    public static string CanonicalPath(string path, string service)
    {
        if (string.IsNullOrEmpty(path)) return "/";
        if (string.Equals(service, "s3", StringComparison.OrdinalIgnoreCase)) return path;
        var parts = path.Split('/');
        var encoded = parts.Select(UriEncodeSegment);
        return string.Join('/', encoded);
    }

    /// <summary>
    /// Percent-encode every byte that isn't in RFC 3986 unreserved
    /// (<c>A-Z a-z 0-9 - . _ ~</c>). Encodes <c>%</c> itself, which
    /// produces SigV4's expected double-encoding when the SDK has
    /// already pre-encoded a path char on the wire.
    /// </summary>
    public static string UriEncodeSegment(string s)
    {
        var bytes = Encoding.UTF8.GetBytes(s);
        var sb = new StringBuilder(bytes.Length);
        foreach (var c in bytes)
        {
            var unreserved =
                (c >= 0x41 && c <= 0x5A) ||
                (c >= 0x61 && c <= 0x7A) ||
                (c >= 0x30 && c <= 0x39) ||
                c == 0x2D || c == 0x2E ||
                c == 0x5F || c == 0x7E;
            if (unreserved) sb.Append((char)c);
            else sb.AppendFormat(CultureInfo.InvariantCulture, "%{0:X2}", c);
        }
        return sb.ToString();
    }

    public static string CanonicalQueryString(string query)
    {
        if (string.IsNullOrEmpty(query)) return "";
        var pairs = query.Split('&').Select(p =>
        {
            var eq = p.IndexOf('=');
            if (eq < 0) return (Key: p, Value: "");
            return (Key: p[..eq], Value: p[(eq + 1)..]);
        }).ToList();
        pairs.Sort((a, b) =>
        {
            var c = string.CompareOrdinal(a.Key, b.Key);
            return c != 0 ? c : string.CompareOrdinal(a.Value, b.Value);
        });
        return string.Join('&', pairs.Select(p => $"{p.Key}={p.Value}"));
    }

    public static string CanonicalHeaderValue(string s)
    {
        var trimmed = s.Trim();
        var sb = new StringBuilder(trimmed.Length);
        var inQuotes = false;
        var lastWasSpace = false;
        foreach (var c in trimmed)
        {
            if (c == '"')
            {
                inQuotes = !inQuotes;
                sb.Append(c);
                lastWasSpace = false;
                continue;
            }
            if (!inQuotes && (c == ' ' || c == '\t'))
            {
                if (!lastWasSpace) { sb.Append(' '); lastWasSpace = true; }
            }
            else
            {
                sb.Append(c);
                lastWasSpace = false;
            }
        }
        return sb.ToString();
    }
}

// Reference: AWS "get-vanilla" SigV4 test (paraphrased)
//
//   Method: GET, Path: /, Query: (empty)
//   Headers: Host: example.amazonaws.com, X-Amz-Date: 20150830T123600Z
//   Body: (empty)
//   AccessKey: AKIDEXAMPLE
//   Secret:    wJalrXUtnFEMI/K7MDENG+bPxRfiCYEXAMPLEKEY
//   Region:    us-east-1, Service: service, Date: 20150830T123600Z
//
// Expected signature: 5fa00fa31553b73ebf1942676e86291e8372ff2a2260956d9b8aae1d763fbf31
// Expected auth:
//   AWS4-HMAC-SHA256
//   Credential=AKIDEXAMPLE/20150830/us-east-1/service/aws4_request,
//   SignedHeaders=host;x-amz-date,
//   Signature=5fa00fa31553b73ebf1942676e86291e8372ff2a2260956d9b8aae1d763fbf31
